src/gpio.rs

```diff
@@ -80,52 +168,157 @@ pub trait GpioPin: SealedPin + Sized + PeripheralType + Into<AnyPin> + 'static {
 }
 
 impl SealedPin for AnyPin {
-    #[inline]
     fn pin_port(&self) -> usize {
         self.port * 32 + self.pin
     }
 
-    #[inline]
     fn gpio(&self) -> &'static crate::pac::gpio0::RegisterBlock {
         self.gpio()
     }
+
+    fn port_reg(&self) -> &'static crate::pac::port0::RegisterBlock {
+        match self.port() {
+            0 => unsafe { &*crate::pac::Port0::ptr() },
+            1 => unsafe { &*crate::pac::Port1::ptr() },
+            2 => unsafe { &*crate::pac::Port2::ptr() },
+            3 => unsafe { &*crate::pac::Port3::ptr() },
+            4 => unsafe { &*crate::pac::Port4::ptr() },
+            _ => panic!("Invalid port: {}", self.port),
+        }
+    }
+
+    fn pcr_reg(&self) -> &'static crate::pac::port0::Pcr0 {
+        let port_reg = self.port_reg();
+        match self.pin() {
+            0 => port_reg.pcr0(),
+            1 => port_reg.pcr1(),
+            2 => port_reg.pcr2(),
+            3 => port_reg.pcr3(),
+            4 => port_reg.pcr4(),
+            5 => port_reg.pcr5(),
+            6 => port_reg.pcr6(),
+            7 => port_reg.pcr7(),
+            8 => port_reg.pcr8(),
+            9 => port_reg.pcr9(),
+            10 => port_reg.pcr10(),
+            11 => port_reg.pcr11(),
+            12 => port_reg.pcr12(),
+            13 => port_reg.pcr13(),
+            14 => port_reg.pcr14(),
+            15 => port_reg.pcr15(),
+            16 => port_reg.pcr16(),
+            17 => port_reg.pcr17(),
+            18 => port_reg.pcr18(),
+            19 => port_reg.pcr19(),
+            20 => port_reg.pcr20(),
+            21 => port_reg.pcr21(),
+            22 => port_reg.pcr22(),
+            23 => port_reg.pcr23(),
+            24 => port_reg.pcr24(),
+            25 => port_reg.pcr25(),
+            26 => port_reg.pcr26(),
+            27 => port_reg.pcr27(),
+            28 => port_reg.pcr28(),
+            29 => port_reg.pcr29(),
+            30 => port_reg.pcr30(),
+            31 => port_reg.pcr31(),
+            _ => panic!("Invalid pin: {}", self.pin),
+        }
+    }
+
+    fn set_function(&self, function: Mux) {
+        self.pcr_reg().modify(|_, w| w.mux().variant(function));
```

Comment:
modify will really read-modify-write, which is not atomic. Usually, we need to wrap modify calls within a `critical_section`. Ditto to all others.

```suggestion
        critical_section::with(|_| self.pcr_reg().modify(|_, w| w.mux().variant(function)));
```

Reply:
Should it be done to all functions, i.e private and public, and of all implementation or only impl SealedPin for AnyPin { as long as the modify is used?

Reply:
actually, please ignore this comment. There's no situation where any of these would race.